### PR TITLE
[SPARK-41195][SQL] Support PIVOT/UNPIVOT with join children

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Today, our SQL parser only supports PIVOT/UNPIVOT at the end of the FROM clause. This is quite limited and it's better to allow PIVOT/UNPIVOT in the join children as well. As a reference, snowflake supports it: https://docs.snowflake.com/en/sql-reference/constructs/from.html

This PR makes PIVOT/UNPIVOT the same level as JOIN. Wherever you can use JOIN to extend a relation, you can also use PIVOT/UNPIVOT. Many SQL syntaxes are supported after this PR
```
FROM t1 PIVOT/UNPIVOT ... JOIN t2  // pivot/unpivot the left table
FROM t1 JOIN t2 PIVOT/UNPIVOT ...  // pivot/unpivot the join result. This is the same before this PR
FROM t1 JOIN (t2 PIVOT/UNPIVOT ...)  // pivot/unpivot the right table
FROM t1 PIVOT/UNPIVOT ... PIVOT/UNPIVOT // nested pivot/unpivot
```

### Why are the changes needed?

make PIVOT/UNPIVOT syntax more flexible

### Does this PR introduce _any_ user-facing change?

Yes, new SQL syntax without any breaking change

### How was this patch tested?

new parser tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @viirya @MaxGekk

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 1]
```diff
@@ -697,12 +697,12 @@ setQuantifier
...
 relation
-    : LATERAL? relationPrimary joinRelation*
+    : LATERAL? relationPrimary pivotClause? unpivotClause? joinRelation*
     ;
 
 joinRelation
-    : (joinType) JOIN LATERAL? right=relationPrimary joinCriteria?
-    | NATURAL joinType JOIN LATERAL? right=relationPrimary
+    : (joinType) JOIN LATERAL? right=relationPrimary joinCriteria? pivotClause? unpivotClause?
+    | NATURAL joinType JOIN LATERAL? right=relationPrimary pivotClause? unpivotClause?
```
Comment (srielau) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:705]: Have you tried simply:
```suggestion
    : (joinType) JOIN LATERAL? right=relationPrimary joinCriteria? pivotClause? unpivotClause?
    | NATURAL joinType JOIN LATERAL? right=relationPrimary
    | pivotClause
    | unpivotClause
```

Also removing the relation entries above?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -1263,60 +1263,71 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
...
+      }
+    }
   }
 
   /**
-   * Join one more [[LogicalPlan]]s to the current logical plan.
+   * Join one more [[LogicalPlan]] to the current logical plan.
    */
-  private def withJoinRelations(base: LogicalPlan, ctx: RelationContext): LogicalPlan = {
-    ctx.joinRelation.asScala.foldLeft(base) { (left, join) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1273]: The actually code change is very small, just remove this loop and rename a few variables.

**Review by gengliangwang — APPROVED** [review]
+1 with the proposal

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Review by viirya — COMMENTED** [review]
Looks good to me.

**Pull request closed without merging**